Card API for pivot tables

robdaemon opened the pull request:
Adds the first /card/:card-id/query endpoint

Commit:
Card API for pivot tables

Adds the first /card/:card-id/query endpoint

Note: this is missing async support, it needs to be added. This change
will break async support in the regular /card API to begin with.

Commit:
broken async stuff

Commit:
Async queries now work

camsaul commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -1,24 +1,30 @@
             [metabase.util.i18n :refer [tru]]
             [schema.core :as s]))
 
+;; TODO: FIX THIS TO WORK WITH ASYNC
+
```
it should work with async now, right?

camsaul reviewed (approved):
LGTM so far.

I think we should figure out whether we should be using some of the fancy QP functions that record the query execution as well, and probably need to make public/embed versions of the endpoint, but I think that can be done as part of the post-release-preview phase

sbelak commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -55,6 +61,47 @@
+                                         (assoc-in [:query :fields] [[:expression "pivot-grouping"]]))
+             all-expected-cols       (qp/query->expected-cols (qp/query->preprocessed col-determination-query))
+             all-queries             (pivot/generate-queries query)
+             first-query             (first all-queries)
```
this could simply be 
`[first-query & rest-queries] (pivot/generate-queries query)`

sbelak commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -55,6 +61,47 @@
+                 ;; shape the row according to the `:column-mapping-fn` above
+                 :row-mapping-fn (fn [row context]
+                                  ;; the first query doesn't need any special mapping, it already has all the columns
+                                   (let [col-mapping (:pivot-column-mapping context)]
```
if-let ?

sbelak commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -55,6 +61,47 @@
+                                       (map (fn [mapping]
+                                              (when mapping
+                                                (nth row mapping)))
+                                            (:pivot-column-mapping context))
```
reuse `col-mapping` var

sbelak commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -55,6 +61,47 @@
+                 ;; this function needs to be executed at the start of every new query to
+                 ;; determine the mapping for maintaining query shape
+                 :column-mapping-fn (fn [query]
+                                      (let [query-cols (map-indexed vector (qp/query->expected-cols (qp/query->preprocessed query)))]
```
Have you considered making `query-cols` a map from name->index? I think it would be more clear than piecing out all the bits from vectors as below.

robdaemon replied:
Maybe? I can give that a shot.

robdaemon commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -55,6 +61,47 @@
+             first-query             (first all-queries)
+             rest-queries            (rest all-queries)]
+         (process-multiple-queries
+          (concat [first-query] rest-queries)
```
heh, good point. I had done something else with them at one point.

Commit:
code review feedback

Commit:
Adding the /public version of the pivot table endpoint

Had to be in a separate endpoint because of how our middleware is structured, but this turns out to be pretty clean looking.
